grouper/fe/handlers/permissions_view.py

```diff
@@ -1,60 +1,49 @@
-from datetime import timedelta
-
+from grouper.entities.pagination import PaginatedList, Pagination
+from grouper.entities.permission import Permission
 from grouper.fe.util import GrouperHandler
-from grouper.user_permissions import user_creatable_permissions
-
+from grouper.repositories.factory import RepositoryFactory
+from grouper.services.factory import ServiceFactory
+from grouper.usecases.factory import UseCaseFactory
+from grouper.usecases.list_permissions import ListPermissionsSortKey, ListPermissionsUI
 
-def _round_timestamp(timestamp):
-    return timestamp - timedelta(seconds=timestamp.second, microseconds=timestamp.microsecond)
 
+class PermissionsView(GrouperHandler, ListPermissionsUI):
+    """Controller for viewing the major permissions list.
 
-class PermissionsView(GrouperHandler):
-    """
-    Controller for viewing the major permissions list. There is no privacy here; the existence of
-    a permission is public.
+    There is no privacy here; the existence of a permission is public.
     """
 
-    def get(self, audited_only=False):
-        offset = int(self.get_argument("offset", 0))
-        limit = int(self.get_argument("limit", 100))
+    def listed_permissions(self, permissions, can_create):
+        # type: (PaginatedList[Permission], bool) -> None
         audited_only = bool(int(self.get_argument("audited", 0)))
-        sort_key = self.get_argument("sort_by", "")
-        sort_dir = self.get_argument("order", "")
-        if limit > 9000:
-            limit = 9000
-
-        sort_keys = {
-            "name": lambda p: p.name,
-            # round timestamps to the nearest minute so permissions created together will
-            # show up alphabetically
-            "date": lambda p: _round_timestamp(p.created_on),
-        }
-
-        if sort_key not in sort_keys:
-            sort_key = "name"
-
-        if sort_dir not in ("asc", "desc"):
-            sort_dir = "asc"
-
-        permissions = sorted(
-            self.graph.get_permissions(audited=audited_only),
-            key=sort_keys[sort_key],
-            reverse=(sort_dir == "desc"),
-        )
-
-        total = len(permissions)
-        permissions = permissions[offset : offset + limit]
-
-        can_create = user_creatable_permissions(self.session, self.current_user)
-
+        sort_key = self.get_argument("sort_by", "name")
```

Comment:
I tend to pass these through the use case as well (so they would be passed back to the UI in the callback), but it seems like Tornado holds onto the state here?

Reply:
Yeah, I thought about doing that but since Tornado stashes all of the state in its handler class, it wasn't necessary. Although on second thought this is a property of the returned list, so I'm going to change that and pass it through since it feels a bit more correct.

Reply:
This unfortunately makes some of the types more complicated, which in turn increases the length of the type signature, which in turn runs afoul of a bug in black. Plumbing the `audited_only` parameter through also requires either a much more complex data structure or (better) splitting the use case. I've therefore left this for now and will take a look in a later diff after there has been a new black release (the bug was fixed on tip but there's been no release to PyPI since).